### PR TITLE
Add a helloworld sample for haskell

### PR DESCRIPTION
**Pull request by abbi-gaurav**
## Proposed Changes

* Use stack as the build tool
* Use scotty as a web framework
* Use docker multi-stage build to create a smaller runtime image

### CODE REVIEW

**Comment by abbi-gaurav** [comment]
I signed the Contributor License Agreement (CLA).

File: serving/samples/helloworld-haskell/app/Main.hs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+import           Data.Monoid        ((<>))
+import           Data.Text.Lazy     (Text)
+import           Data.Text.Lazy
+import           System.Environment (lookupEnv)
+import           Web.Scotty         (ActionM, ScottyM, scotty)
+import           Web.Scotty.Trans
+
+main :: IO ()
+main = do
+  m <- lookupEnv "TARGET"
```
Comment (ian-mi) [inline serving/samples/helloworld-haskell/app/Main.hs]: Seems clearer without the intermediate value:
```t <- fromMaybe "NOT SPECIFIED" <$> lookupEnv "TARGET"```

**Comment by abbi-gaurav** [comment]
@ian-mi  I have done the changes proposed by you. Please have a look.

**Comment by mattmoor** [comment]
/approve

will let @ian-mi give `/lgtm`

File: serving/samples/helloworld-haskell/README.md [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    ```
+
+1. Now that your service is created, Knative will perform the following steps:
+   * Create a new immutable revision for this version of the app.
+   * Network programming to create a route, ingress, service, and load balance for your app.
+   * Automatically scale your pods up and down (including to zero active pods).
+
+1. To find the IP address for your service, enter
+   `kubectl get svc knative-ingressgateway -n istio-system` to get the ingress IP for your
+   cluster. If your cluster is new, it may take sometime for the service to get asssigned
```
[the remark this answers is not in the document]
Comment (abbi-gaurav) [reply serving/samples/helloworld-haskell/README.md]: @jonjohnsonjr Thanks for observing and providing feedback. I have done the changes.

**Comment by abbi-gaurav** [comment]
@samodell I have updated the changes suggested for README.md. Due to this `lgtm` label has been removed. Please let me know if any further action is required from me.

**Comment by abbi-gaurav** [comment]
@samodell I have assigned the PR to you for the necessary edits.

**Comment by samodell** [comment]
@abbi-gaurav  Working through this today!

File: serving/samples/helloworld-haskell/README.md [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+Once you have recreated the sample code files (or used the files in the sample
+folder) you're ready to build and deploy the sample app.
+
+1. Use Docker to build the sample code into a container. To build and push with
+   Docker Hub, enter these commands replacing `{username}` with your
+   Docker Hub username:
+
+    ```shell
+    # Build the container on your local machine
+    docker build -t {username}/helloworld-haskell .
```
Comment (samodell) [inline serving/samples/helloworld-haskell/README.md]: @abbi-gaurav 

I'm getting an error while building my container locally; here's the output:

Step 5/10 : RUN stack build --copy-bins
 ---> Running in 06e32c35baf3
Updating package index Hackage (mirrored at https://s3.amazonaws.com/hackage.fpcomplete.com/) ...
Selected mirror https://s3.amazonaws.com/hackage.fpcomplete.com/
Downloading root
Selected mirror https://s3.amazonaws.com/hackage.fpcomplete.com/
Downloading timestamp
Downloading snapshot
Downloading mirrors
Cannot update index (no local copy)
Downloading index
Updated package index downloaded
Update complete
Populating index cache ...
Populated index cache.
Error parsing targets: The specified targets matched no packages.
Perhaps you need to run 'stack init'?
The command '/bin/sh -c stack build --copy-bins' returned a non-zero code: 1

Any suggestions? I believe I've copied the sample files correctly.

Comment (abbi-gaurav) [reply serving/samples/helloworld-haskell/README.md]: @samodell let me have a look. I will get back to you soon.

Comment (abbi-gaurav) [reply serving/samples/helloworld-haskell/README.md]: @samodell I do not see the error when I run the command. However, let me check if adding `stack init` as suggested will help.

```bash
$ docker build -t gabbi/helloworld-haskell .
Sending build context to Docker daemon  9.219MB
Step 1/10 : FROM haskell:8.2.2 as builder
8.2.2: Pulling from library/haskell
55cbf04beb70: Pull complete
18a5c749e623: Pull complete
Digest: sha256:5bebc63cf629a42e66e87d26168dac1f82c41467e881c95220852f3e4dbbfda8
Status: Downloaded newer image for haskell:8.2.2
 ---> 0cb8be425b81
Step 2/10 : WORKDIR /app
 ---> Running in 381568b9f3aa
Removing intermediate container 381568b9f3aa
 ---> 6f96362fe9e5
Step 3/10 : ADD . /app
 ---> 5d9b40453b56
Step 4/10 : RUN stack setup
 ---> Running in ea4d4477f5f4
Downloading lts-10.7 build plan ...
Downloaded lts-10.7 build plan.
stack will use the GHC on your PATH
For more information on paths, see 'stack path' and 'stack exec env'
To use this GHC and packages outside of a project, consider using:
stack ghc, stack ghci, stack runghc, or stack exec
Removing intermediate container ea4d4477f5f4
 ---> 191b0e83fe56
Step 5/10 : RUN stack build --copy-bins
 ---> Running in bad6eb0cf548
Updating package index Hackage (mirrored at https://s3.amazonaws.com/hackage.fpcomplete.com/) ...
Selected mirror https://s3.amazonaws.com/hackage.fpcomplete.com/
Downloading root
Selected mirror https://s3.amazonaws.com/hackage.fpcomplete.com/
Downloading timestamp
Downloading snapshot
Downloading mirrors
Cannot update index (no local copy)
Downloading index
Updated package index downloaded
Update complete
Populating index cache ...
Populated index cache.
[1 of 2] Compiling Main             ( /root/.stack/setup-exe-src/setup-mPHDZzAJ.hs, /root/.stack/setup-exe-src/setup-mPHDZzAJ.o )
[2 of 2] Compiling StackSetupShim   ( /root/.stack/setup-exe-src/setup-shim-mPHDZzAJ.hs, /root/.stack/setup-exe-src/setup-shim-mPHDZzAJ.o )
Linking /root/.stack/setup-exe-cache/x86_64-linux/tmp-Cabal-simple_mPHDZzAJ_2.0.1.0_ghc-8.2.2 ...
base-compat-0.9.3: download
appar-0.1.4: download
```

Comment (samodell) [reply serving/samples/helloworld-haskell/README.md]: @abbi-gaurav  Oh, I re-checked my file structure and found I made the Dockerfile in the wrong location. My mistake; the build is working fine now. I'll continue walking through the guide and get back to you with edits by EOD PST.

Comment (abbi-gaurav) [reply serving/samples/helloworld-haskell/README.md]: @samodell I confirmed that we do not need to do any changes and steps are working fine. 

Could you please try running the `docker build` command from the directory in the repository `docs/serving/samples/helloworld-haskell`?

File: serving/samples/helloworld-haskell/README.md [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    runLatest:
+        configuration:
+        revisionTemplate:
+            spec:
+            container:
+                image: docker.io/{username}/helloworld-haskell
+                env:
+                - name: TARGET
+                value: "Haskell Sample v1"
+    ```
```
Comment (samodell) [inline serving/samples/helloworld-haskell/README.md]: The indentation here is off. Copying and pasting from here, I couldn't apply the service.yaml without getting an error. When I went and copied the raw version of the file, the service was successfully created.

I would re-copy from your service.yaml file to get the indentation right.

File: serving/samples/helloworld-haskell/README.md [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+
+## Prerequisites
+
+* A Kubernetes cluster with Knative installed. Follow the
+  [installation instructions](https://github.com/knative/docs/blob/master/install/README.md) if you need
+  to create one.
+* [Docker](https://www.docker.com) installed and running on your local machine,
+  and a Docker Hub account configured (we'll use it for a container registry).
+
+## Steps to recreate the sample code
```
Comment (samodell) [inline serving/samples/helloworld-haskell/README.md]: Change header to "Recreating the sample code"

File: serving/samples/helloworld-haskell/README.md (hunk 3 above)
Comment (samodell) [reply serving/samples/helloworld-haskell/README.md]: Was able to get through and successfully deploy.

File: serving/samples/helloworld-haskell/README.md (hunk 4 above)
Comment (abbi-gaurav) [reply serving/samples/helloworld-haskell/README.md]: I will work on this today. Thanks for the inputs.

File: serving/samples/helloworld-haskell/README.md (hunk 5 above)
Comment (abbi-gaurav) [reply serving/samples/helloworld-haskell/README.md]: I will work on this today. Thanks for the inputs.

**Comment by abbi-gaurav** [comment]
@samodell I have applied the suggested changes in the commit. https://github.com/knative/docs/pull/316/commits/9f057e9f2a4c6a9aacf6aa5db7cf329b6ac3ddbe

Please let me know if anything else required.

**Pull request merged**